[Step13 error / Step 14-16 prepare] Step 13 部署遇到問題，並新增後續步驟的一些實作

snowmint opened the pull request:
[x]步驟13: 網站佈署
    遇到問題花了一些時間仍無法解決，因此整理至：https://stackoverflow.com/questions/68163849/deploy-ruby-on-rails-project-on-heroku-failed-you-have-deleted-from-the-gemfil

並繼續加入後續步驟，發現其中有一些已經在前面或過程中實作過

步驟14: 加入結束時間
步驟15: 加入狀態，並且能夠查詢
    在任務上加入狀態（待處理、進行中、完成）
    在列表頁面，要能夠以標題和狀態進行查詢
步驟16: 設定優先順序
    在任務上加入優先順序（高、中、低）
    列表頁可依優先順序做排序

![Search](https://user-images.githubusercontent.com/7868828/123647130-ac64db80-d7dc-11eb-917f-a007c0639fbe.jpg)

snowmint committed:
[step 15 prepare] Add search

snowmint committed:
[Step 15 prepare] Add Search function with task_name, priority, and status

snowmint committed:
Bundler fix

snowmint committed:
commiting Gemfile.lock

snowmint committed:
commiting Gemfile.lock

ryudoawaru commented on app/models/task/filter.rb:
```diff
@@ -0,0 +1,19 @@
+class Task::Filter
+
+    def filter(scope, query_params)
```
請問這個是基於什麼想法這樣設計的？

snowmint replied:
這邊是因為想要達到 task 名稱跟優先度等多個屬性可以同時選擇並搜尋。

ryudoawaru replied:
那寫在 model scope 或是寫 class method 就好了？

snowmint replied:
已經修改好了，移至 task_controller.rb

ryudoawaru commented on app/models/task.rb:
```diff
@@ -4,9 +4,15 @@ class Task < ApplicationRecord
   validates :user_id, :start_time, :end_time, :task_name, presence: true
-  validates :priority, presence: true, numericality: {less_than_or_equal_to: 10}
-  validates :status, presence: true, numericality: {less_than_or_equal_to: 6}
+  validates :priority, presence: true, numericality: {#['0', 'Urgent'], ['1', 'Highest'], ['2', 'High'], ['3', 'Normal'], ['4', 'Low'], ['5', Lowest]
```
非必要 comment 不要留下來

snowmint replied:
好的知道了。

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -70,11 +71,24 @@ def destroy
     end
+
+    def search  
+      if params[:search].blank?  
```
如果沒有特別的需求，我們會用 index 做 search 的事，通常不會兩組 action 和 view

snowmint replied:
這邊是遺留下來的 code 我把它移除，因為在做搜尋的時候有做了第一個版本是只能以關鍵字搜尋的，但覺得不符合需求因此改成用上面 filter.rb 的方法，只用 index 一個 view

snowmint replied:
每個屬性篩選條件都可填或不填
![Screenshot from 2021-06-28 21-41-36](https://user-images.githubusercontent.com/7868828/123827506-8828fd80-d8b5-11eb-8ba9-22dcbea11d96.png)
![Screenshot from 2021-06-28 21-41-44](https://user-images.githubusercontent.com/7868828/123827522-8a8b5780-d8b5-11eb-9c30-f47f06edc6a8.png)
![Screenshot from 2021-06-28 21-41-51](https://user-images.githubusercontent.com/7868828/123827074-2b2d4780-d8b5-11eb-9718-5fc3e67d7d72.png)
![Screenshot from 2021-06-28 21-41-59](https://user-images.githubusercontent.com/7868828/123827538-8e1ede80-d8b5-11eb-9014-cbf5dda6b074.png)

ryudoawaru commented on app/views/tasks/search.html.erb:
```diff
@@ -0,0 +1,30 @@
+<table class="table table-striped table-hover">
+  <thead>
+    <tr>
+      <th> <%= raw(t('activerecord.attributes.tasks.user_id')) %> </th>
```
請善用 [human_attributes_name](http://api.rubyonrails.org/classes/ActiveModel/Translation.html#method-i-human_attribute_name)

snowmint committed:
[Modify] modify last pull request

ryudoawaru commented:
那個這包變 5000 多檔案
一定是有出了啥問題
因為我打不太開 files changed（會卡住）所以沒辦法直接看情況
請你先處理這個問題
![截圖 2021-06-30 上午12 13 57](https://user-images.githubusercontent.com/212290/123832585-1219aa80-d938-11eb-8472-1ce7ed972e86.jpg)

snowmint commented:
> 那個這包變 5000 多檔案
> 一定是有出了啥問題
> 因為我打不太開 files changed（會卡住）所以沒辦法直接看情況
> 請你先處理這個問題
> ![截圖 2021-06-30 上午12 13 57](https://user-images.githubusercontent.com/212290/123832585-1219aa80-d938-11eb-8472-1ce7ed972e86.jpg)


我先用瀏覽器的 loading 空檔取消大部分檔案的顯示，發現是 /vendor/bundle/ruby/3.0.0/gems 下多出大量的檔案

或許是之前為了部署到 Heroku 上我多次下了 bundle 相關的指令
bundle update
git add .
git push "Gemfile.lock"
git push heroku master

rm -rf ~/.bundle/ ~/.gem/ .bundle/ Gemfile.lock
bundle install
git push heroku master

造成 gemfile 大量更動

![螢幕擷取畫面 (2)](https://user-images.githubusercontent.com/7868828/123838719-24a4cd00-d8c1-11eb-8b0b-5a308ec8e85f.png)

snowmint commented:
我真正有更動到的檔案是以下這些
![螢幕擷取畫面 2021-06-30 012428](https://user-images.githubusercontent.com/7868828/123841330-42bffc80-d8c4-11eb-888b-4846d71b8f39.png)

snowmint committed:
[Modify] human_attribute_name update

ryudoawaru commented:
那個下面的檔案本來就不該進到 Repo 的

snowmint committed:
[Remove/Ignore] /vendor/bundle

snowmint commented:
@ryudoawaru 

將 /vendor/bundle 移除並加入 .gitignore 中

snowmint committed:
[Modify] Filter move into task_controller

snowmint commented:
Heroku 部署不上去的部份目前還未找到解法
嘗試過的解法︰
1.
```
bundle update
git add .
git push "Gemfile.lock"
git push heroku master
```
2.
```
rm -rf ~/.bundle/ ~/.gem/ .bundle/ Gemfile.lock
bundle install
git push heroku master
```
官方建議的解法︰
```
bundle lock --add-platform ruby
$ bundle lock --add-platform x86_64-linux
$ bundle install
$ git add . ; git commit -m "Bundler fix"
```
但仍然 push 失敗

Build log:
```
-----> Building on the Heroku-20 stack
-----> Determining which buildpack to use for this app
 !     Warning: Multiple default buildpacks reported the ability to handle this app. The first buildpack in the list below will be used.
			Detected buildpacks: Ruby,Node.js
			See https://devcenter.heroku.com/articles/buildpacks#buildpack-detect-order
-----> Ruby app detected
-----> Installing bundler 2.2.21
-----> Removing BUNDLED WITH version in the Gemfile.lock
-----> Compiling Ruby/Rails
-----> Using Ruby version: ruby-3.0.1
-----> Installing dependencies using bundler 2.2.21
       Running: BUNDLE_WITHOUT='development:test' BUNDLE_PATH=vendor/bundle BUNDLE_BIN=vendor/bundle/bin BUNDLE_DEPLOYMENT=1 bundle install -j4
       You are trying to install in deployment mode after changing
       your Gemfile. Run 'bundle install' elsewhere and add the
       updated Gemfile.lock to version control.
       
       If this is a development machine, remove the /tmp/build_c6be2a0c/Gemfile freeze 
       by running ''.
       
       You have deleted from the Gemfile:
       * pg
       Bundler Output: You are trying to install in deployment mode after changing
       your Gemfile. Run 'bundle install' elsewhere and add the
       updated Gemfile.lock to version control.
       
       If this is a development machine, remove the /tmp/build_c6be2a0c/Gemfile freeze 
       by running ''.

       You have deleted from the Gemfile:
       * pg

 !
 !     Failed to install gems via Bundler.
 !
 !     Push rejected, failed to compile Ruby app.
 !     Push failed
```

ryudoawaru commented on app/controllers/tasks_controller.rb:
```diff
@@ -70,11 +71,33 @@ def destroy
     end
-  
+
+    class Task::Filter
```
我前面的意思是請你把 scope 寫在 model 裡的 scope 或是寫成 model 的 class method（其實跟 scope 一樣）

https://api.rubyonrails.org/v6.1.3.2/classes/ActiveRecord/Scoping/Named/ClassMethods.html

因為這個查詢並不複雜，還沒有必要為它去特別開一個 class 弄

然後把 class 宣告放在 controller 是有點奇怪的，一般來說 controller 要盡量照 restful，每個 method 的 code 要盡量以引用 model 寫好的邏輯去 CRUD 資料為主

snowmint replied:
試過用 AND 來 query ，但不知道如何能夠在單一一個選項為空的時候查詢
```
scope :filtered, ->(query_params) {
    where("task_name ILIKE :search AND priority = :priority AND status = :status", search: "%#{query_params[:keyword]}%", priority: query_params[:priority], status: query_params[:status])
 }
```
使用 class method 也無法使用 if 條件來區分條件

ryudoawaru replied:
query_params 是 Hash 吧

`keyword, priority, status = query_params.values_at(:keyword, :priority, :status)`

把三個分別做個別的 scope

ryudoawaru replied:
scope 裡 return nil 或 all 就表示 skip 這個 scope（不會加上任何條件）

可以朝這方向想

snowmint replied:
目前改成以下這樣，但是 select 中 include_blank = true 會回傳 '' 而不是 nil，所以回傳都會以 blank 去搜尋而報錯

```
  scope :status, -> (query_params) { where(status: query_params[:status]) }
  scope :priority, -> (query_params) { where(priority: query_params[:priority]) }
  scope :keyword, -> (query_params) { where("task_name ILIKE :search", search: "%#{query_params[:keyword]}%") }

  def self.filtered(query_params)
    keyword, priority, status = query_params.values_at(:keyword, :priority, :status)
    Task.where("task_name ILIKE :search AND priority = :priority AND status = :status", search: "%#{keyword}%", priority: priority, status: status)
  end
```
![Screenshot from 2021-07-01 17-46-27](https://user-images.githubusercontent.com/7868828/124112348-0fde4b80-da1f-11eb-9964-08b3466ec35d.png)

snowmint replied:
也嘗試將 select form 的 list blank 對應到 nil，但依然會在未選擇時報錯

```
  <%= form_for :query, url: root_path, method: :get do |form| %>
    <div class="field form-group" style="display:inline-block;">
      <%= form.text_field :keyword, placeholder: raw(t('task_relate.enter_keyword')), value: params.dig('query', 'keyword'), class: "form-control" %>
    </div>

    <div class="field form-group" style="display:inline-block;">
      <%= form.label Task.human_attribute_name("tasks.priority"), for: "task_priority" %>
      <%= form.select :priority, [['', nil], ['Urgent', 0], ['Highest', 1], ['High', 2], ['Normal', 3], ['Low', 4], ['Lowest', 5]], selected: params.dig('query', 'priority'), class: "form-select form-control", id: "task_priority"%>
    </div>

    <div class="field form-group" style="display:inline-block;">
      <%= form.label Task.human_attribute_name("tasks.status"), for: "task_status" %>
      <%= form.select :status, [['', nil], ['Pending', 0], ['In progress', 1], ['Complete', 2]], selected: params.dig('query', 'status'), class: "form-select form-control", id: "task_status" %>
    </div>

    <div class="actions form-group" style="display:inline-block;">
      <%= form.submit raw(t('task_relate.serch_task')), class: "form-control" %>
    </div>
  <% end  %>
```

如果在初始 NULL 的情況下會顯示空 list，但如果經選擇空白則還是會變更為 '' (blank) 就會報錯
 ![Screenshot from 2021-07-01 19-45-22](https://user-images.githubusercontent.com/7868828/124119343-3ef8bb00-da27-11eb-807f-2ea424425085.png)

需要 select 兩個都選擇才能正常執行
![Screenshot from 2021-07-01 19-49-07](https://user-images.githubusercontent.com/7868828/124119726-b4648b80-da27-11eb-9307-ce9b1921a896.png)

ryudoawaru replied:
```ruby 
# in Task model
scope :filter_by_status, -> (status) { where(status: status) if status }
# priority / keyword 比照辦理

# 然後在 controller

keyword, priority, status = params.values_at(:keyword, :priority, :status)

@tasks = Task.filter_by_status(status).filter_by_keyword(keyword).filter_by_priority(priority)

```

ryudoawaru commented:
deploy 問題請放群組問

snowmint committed:
[Modify] Use Class method to search

snowmint commented:
在看到留言之前我已經有 push 一版是用 class method 達到 search 了，不知道可不可以？
```
def self.by_status(query_params)
    if query_params[:status].present?
      where(status: query_params[:status])
    else
      all
    end
  end

  def self.by_priority(query_params)
    if (query_params[:priority].present?)
      where(priority: query_params[:priority])
    else
      all
    end
  end

  def self.by_keyword(query_params)
    if query_params[:keyword].present?
      where("task_name ILIKE :search", search: "%#{query_params[:keyword]}%")
    else
      all
    end
  end

  def self.filtered(query_params)
    #keyword, priority, status = query_params.values_at(:keyword, :priority, :status)
    #Task.where("task_name ILIKE :search AND priority = :priority AND status = :status", search: "%#{keyword}%", priority: priority, status: status)
    #Task.where("task_name ILIKE :search AND priority = :priority AND status = :status", search: "%#{query_params[:keyword]}%", priority: query_params[:priority], status: query_params[:status])
    Task.by_status(query_params).by_priority(query_params).by_keyword(query_params)
  end
```

ryudoawaru commented on app/models/task.rb:
```diff
@@ -4,9 +4,55 @@ class Task < ApplicationRecord
+
+  def self.by_priority(query_params)
+    if (query_params[:priority].present?)
+      where(priority: query_params[:priority])
```
你這個 scope 只是查一個 priority，不需要帶整個 hash，會顯的多餘

ryudoawaru commented:
剩一個問題其它OK

如果等下解的話可直接 merge

snowmint committed:
[Modify] pass parameter with single one

snowmint commented:
僅傳單一參數

```
  def self.by_status(status)
    if status.present?
      where(status: status)
    else
      all
    end
  end

  def self.by_priority(priority)
    if (priority.present?)
      where(priority: priority)
    else
      all
    end
  end

  def self.by_keyword(keyword)
    if keyword.present?
      where("task_name ILIKE :search", search: "%#{keyword}%")
    else
      all
    end
  end

  def self.filtered(query_params)
    Task.by_status(query_params[:status]).by_priority(query_params[:priority]).by_keyword(query_params[:keyword])
  end
```
